File: java/lib/src/main/java/algorithm/sorting/Sort.java
package algorithm.sorting;

public abstract class Sort {
  public boolean compare(int x, int y, SortType type) {
    return type == SortType.asc ? x > y : x < y;
  }

  public void swap(int[] arr, int i, int j) {
    int tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
  }

  protected abstract int[] sort(int[] arr, SortType type);

}

File: kotlin/lib/src/main/kotlin/algorithm/sorting/Sort.kt
package algorithm.sorting

fun Array<Int>.swap(i: Int, j: Int) {
  val tmp = this[i]
  this[i] = this[j]
  this[j] = tmp
}

enum class SortType {
  asc,
  desc
}

public abstract class Sort {
  protected fun compare(x: Int, y: Int, type: SortType): Boolean {
    return if (type == SortType.asc) x > y else x < y
  }
  abstract fun sort(arr: Array<Int>, type: SortType = SortType.asc): Array<Int>
}
